include/deal.II/base/mpi_noncontiguous_vector.h

```diff
@@ -0,0 +1,256 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2020 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_mpi_noncontiguous_vector_h
+#define dealii_mpi_noncontiguous_vector_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/mpi.h>
+#include <deal.II/base/mpi.templates.h>
```

Comment:
Nope. I need `Utilities::MPI::internal::mpi_type_id()`.

Reply:
That's a pity - I'm always a bit uncomfortable with including definitions which are meant to be compilation targets in header files because of the duplicate symbols this would lead to. Now here it should work because the functions are all templates so no symbols are generated without instantiations. We try to be conservative with including other `.templates.h` files in headers - it is mostly the various `_accessor.templates.h` and `_iterator.templates.h` that come to mind. Now, this is not the topic of this PR, so we just keep it this way.

Reply:
How do you like it this way?

https://github.com/dealii/dealii/blob/3778ed08040b39a4858b6f1a85449e29d70b987f/source/base/mpi_noncontiguous_partitioner.inst.in#L26-L49

This way, I was able to move the include to the .cc-file.

If you are happy, I will squash the commits.